02_cancer_type_classification/nbconverted/lasso_range_all.py

```diff
@@ -0,0 +1,155 @@
+        )
+        
+nz_coefs_df = pd.DataFrame(
+    nz_coefs_df,
+    columns=['gene', 'cancer_type', 'lasso_param', 'seed', 'fold', 'nz_coefs']
+)
+nz_coefs_df.head()
+
+
+# ### Get performance information for each lasso penalty
+
+# In[4]:
+
+
+perf_df = au.load_prediction_results_lasso_range(results_dir, training_dataset)
+perf_df.head()
+
+
+# ### Compare feature selection with performance
+
+# In[5]:
+
+
+coefs_perf_df = (nz_coefs_df
+    .rename(columns={'cancer_type': 'holdout_cancer_type'})
+    .merge(perf_df[perf_df.signal == 'signal'],
+           on=['gene', 'holdout_cancer_type', 'seed', 'fold', 'lasso_param'])
+    .drop(columns=['signal'])
+)
+
+coefs_perf_df.head()
+
+
+# In[6]:
+
+
+sns.set({'figure.figsize': (8, 6)})
+
+sns.histplot(coefs_perf_df.nz_coefs, log_scale=(True, False))
+plt.gca().axvline(coefs_perf_df.nz_coefs.quantile(q=0.1), linestyle='--')
+plt.title('Distribution of feature count across cancer types/folds')
+plt.xlabel('Number of nonzero features')
+
+coefs_perf_df.loc[coefs_perf_df.nz_coefs.sort_values()[:8].index, :]
+
+
+# In[7]:
+
+
+# look at correlation for each cancer type individually
+# positive correlation => more features, better performance
+corr_cancer_type_df = []
+nz_coefs_cutoff = coefs_perf_df.nz_coefs.quantile(q=0.1)
+
+if nz_coefs_cutoff is not None:
+    coefs_perf_df = coefs_perf_df[coefs_perf_df.nz_coefs > nz_coefs_cutoff].copy()
+
+for gene in coefs_perf_df.gene.unique():
+    for cancer_type in coefs_perf_df.holdout_cancer_type.unique():
+        corr_df = coefs_perf_df[
+            (coefs_perf_df.gene == gene) &
+            (coefs_perf_df.holdout_cancer_type == cancer_type) &
+            (coefs_perf_df.data_type == 'test')
+        ]
+        try:
+            r, p = pearsonr(corr_df.nz_coefs.values, corr_df.aupr.values)
+        except ValueError:
+            # this happens when the model wasn't trained on the cancer type
+            # for the given gene, just skip
+            continue
+        corr_cancer_type_df.append(
+            [gene, cancer_type, r, p]
+        )
+    
+corr_cancer_type_df = pd.DataFrame(
+    corr_cancer_type_df,
+    columns=['gene', 'cancer_type', 'pearson_r', 'pearson_pval']
+).sort_values(by='pearson_r', ascending=False)
+
+print(corr_cancer_type_df.shape)
+corr_cancer_type_df.head()
+
+
+# In[8]:
+
+
+# plot test performance vs. number of nonzero features
+sns.set({'figure.figsize': (12, 6)})
+
+ax = sns.boxplot(data=corr_cancer_type_df, x='gene', y='pearson_r')
+ax.axhline(0.0, linestyle='--', color='black')
+plt.title(f'Model size/performance correlations across cancer types, per gene (nonzero cutoff: {nz_coefs_cutoff})')
+plt.xlabel('Gene')
+plt.ylabel('Pearson correlation')
+
+
+# This box plot suggests that across these 4 genes, there tends to be a positive correlation between the number of features in the model and generalization performance (i.e. in general, more features in the model => better generalization). 
```

Comment:
Correlations can be tricky. Maybe plot 1-2 non aggregated to make sure it look like there is a true positive correlation between # of features and performance?

Reply:
Just realized I should have looked at lasso_range_gene first. Look like you did this already.